arch/xtensa/src/esp32/esp32_wlan.c

```diff
@@ -0,0 +1,1589 @@
+
+  DEBUGASSERT(priv->esp_dev.d_buf != NULL);
+
+  /* If the polling resulted in data that should be sent out on the network,
+   * the field d_len is set to a value > 0.
+   */
+
+  if (priv->esp_dev.d_len > 0)
+    {
+      /* Look up the destination MAC address and add it to the Ethernet
+       * header.
+       */
+
+#ifdef CONFIG_NET_IPv4
+#ifdef CONFIG_NET_IPv6
+      if (IFF_IS_IPv4(priv->esp_dev.d_flags))
+#endif
+        {
+          arp_out(&priv->esp_dev);
+        }
+#endif /* CONFIG_NET_IPv4 */
+
+#ifdef CONFIG_NET_IPv6
+#ifdef CONFIG_NET_IPv4
+      else
+#endif
+        {
+          neighbor_out(&priv->esp_dev);
+        }
+#endif /* CONFIG_NET_IPv6 */
+
+      if (!devif_loopback(&priv->esp_dev))
+        {
+          /* Send the packet */
+
+          int ret = esp_transmit(priv);
+          if (ret != OK)
+            {
+              wlerr("TX failed\r\n");
+              return -EBUSY;
+            }
+        }
+    }
+
+  /* If zero is returned, the polling will continue until
+   * all connections have been examined.
+   */
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_rxpoll
+ *
+ * Description:
+ *   Process RX frames
+ *
+ * Input Parameters:
+ *   arg - context of device to use
+ *
+ * Returned Value:
+ *   OK on success
+ *
+ * Assumptions:
+ *   The network is locked.
+ *
+ ****************************************************************************/
+
+static void esp_rxpoll(FAR void *arg)
+{
+  FAR struct esp_dev_s *priv = (FAR struct esp_dev_s *)arg;
+
+  if (priv->esp_dev.d_buf == NULL)
+    {
+      priv->esp_dev.d_buf = priv->rxbuf;
+      priv->esp_dev.d_len = priv->rx_len;
+    }
+  else
+    {
+      wlinfo("priv->esp_dev.d_buf != NULL");
+      return;
+    }
+
+  /* Lock the network and serialize driver operations if necessary.
+   * NOTE: Serialization is only required in the case where the driver work
+   * is performed on an LP worker thread and where more than one LP worker
+   * thread has been configured.
+   */
+
+  esp_takesem(priv);
+
+  net_lock();
+
+  esp_receive(priv);
+
+  /* Check if a packet transmission just completed.  If so, call esp_txdone.
+   * This may disable further Tx interrupts if there are no pending
+   * transmissions.
+   */
+
+  net_unlock();
+  if (priv->esp_dev.d_buf)
+    {
+      priv->esp_dev.d_buf = NULL;
+      memset(priv->rxbuf, 0x0, sizeof(priv->rxbuf));
+      priv->rx_len = 0;
+    }
+
+  esp_givesem(priv);
+}
+
+/****************************************************************************
+ * Function: esp_dopoll
+ *
+ * Description:
+ *   The function is called in order to perform an out-of-sequence TX poll.
+ *   This is done:
+ *
+ *   1. After completion of a transmission (esp32_txdone),
+ *   2. When new TX data is available (esp32_txavail_process), and
+ *   3. After a TX timeout to restart the sending process
+ *      (stm32_txtimeout_process).
```

Comment:
I've found that these steps are replicated through several ethernet driver implementations (STM32, SAM, tiva, rx65n...), and all of them are reference the same functions that no longer exist (e.g. xxx_txavail_process).

Reply:
In this case it's referring to the `esp_txavail` function.
Fixed too. Thanks.